libs/sdk-core/src/breez_services.rs

```diff
@@ -930,6 +920,109 @@ impl BreezServices {
+            max_sat: fee_info.max,
+            fees_hash: fee_info.fees_hash,
+        })
+    }
+
+    /// Supersedes [BreezServices::fetch_reverse_swap_fees]
+    ///
+    /// ### Errors
+    ///
+    /// - `FeePromiseChanged`: This indicates the fee promise used in the argument is no longer valid.
+    /// Please use `fetch_onchain_limits` to get the new fee promise. Note that the returned limits
+    /// (`min_sat` and `max_sat`) may have also changed, so make sure to use these new values when
+    /// validating user input.
+    ///
+    /// - `OutOfRange`: This indicates the send amount is outside the range of minimum and maximum
+    /// values returned by `fetch_onchain_limits`.
+    pub async fn prepare_onchain_payment(
+        &self,
+        req: PrepareOnchainPaymentRequest,
+    ) -> Result<PrepareOnchainPaymentResponse, PayOnchainError> {
+        let fee_info = self.btc_send_swapper.fetch_reverse_swap_fees().await?;
+
+        // Calculate (send_amt, recv_amt) from the inputs and fees
+        let fees_lockup = fee_info.fees_lockup;
+        let p = fee_info.fees_percentage;
+        let fees_claim = BTCSendSwap::calculate_claim_tx_fee(req.claim_tx_feerate)?;
+        let (send_amt, recv_amt) = match req.amount_type {
+            SwapAmountType::Send => {
+                let temp_send_amt = req.amount_sat;
+                let service_fees = ((temp_send_amt as f64) * p / 100.0) as u64;
+                let total_fees = service_fees + fees_lockup + fees_claim;
+                ensure_sdk!(
+                    temp_send_amt > total_fees,
+                    PayOnchainError::generic(
+                        "Send amount is not high enough to account for all fees"
+                    )
+                );
+
+                (temp_send_amt, temp_send_amt - total_fees)
+            }
+            SwapAmountType::Receive => {
+                let temp_recv_amt = req.amount_sat;
+                let send_amt_and_service_fee = temp_recv_amt + fees_lockup + fees_claim;
+                let temp_send_amt = send_amt_and_service_fee as f64 * 100.0 / (100.0 - p);
+
+                (temp_send_amt as u64, temp_recv_amt)
+            }
+        };
+
+        let temp_res = PrepareOnchainPaymentResponse {
+            fees_hash: fee_info.fees_hash.clone(),
+            fees_percentage: p,
+            fees_lockup,
+            fees_claim,
+            send_amount_sat: send_amt,
+            receive_amount_sat: recv_amt,
+            total_fees: send_amt - recv_amt,
+        };
+
+        temp_res
+            .validate_against_fresh_fees_info(fee_info)
+            .map(|valid| valid.0)
+    }
+
+    /// Creates a reverse swap and attempts to pay the HODL invoice
+    ///
+    /// Supersedes [BreezServices::send_onchain]
+    pub async fn pay_onchain(
+        &self,
+        req: PayOnchainRequest,
+    ) -> Result<PayOnchainResponse, PayOnchainError> {
+        let fee_info = self.btc_send_swapper.fetch_reverse_swap_fees().await?;
```

Comment:
Why do we need to fetch the fees again if we have the fee_hash? Isn't it better to create the swap using this fee hash which is promised to return the same fees the user got from the prepare call? We can then validate against the returned swap.

Reply:
Done: 6c876a6d99d3160d0a85f688e0e74eca734cef4f